### PR TITLE
Allow types to the left of `:` in where predicates

### PR DESCRIPTION
**Pull request by regiontog**
Adresses #584

### CODE REVIEW

File: crates/ra_syntax/src/grammar/type_params.rs [hunk 1]
```diff
@@ -104,22 +104,32 @@ pub(super) fn opt_where_clause(p: &mut Parser) {
     }
     let m = p.start();
     p.bump();
-    loop {
-        if !(paths::is_path_start(p)
-            || p.current() == LIFETIME
-            || p.current() == FOR_KW
-            || p.current() == L_ANGLE)
-        {
-            break;
```
Comment (regiontog) [inline crates/ra_syntax/src/grammar/type_params.rs]: Is it necessary to break if it is not a lifetime or type here?

Comment (matklad) [reply crates/ra_syntax/src/grammar/type_params.rs]: Yep, break is pretty important here for the error recovery. For the code like

```
fn foo()
where

struct Foo {}
```

we should end where clause when we see `struct` keyword, and that's what this `break` achieves.  Checking for `p.current() == L_CURLY || p.current() == SEMI || p.current() == EQ` will work for valid code correctly, but would be suboptimal for syntactically invalid code, which we need to handle as well.

I think `is_where_clause_end` should look like `!(p.at(LIFETIME) || p.at_ts(TYPE_FIRST) && !p.at(IMPL_KW))`. `TYPE_FIRST` is a set of tokens which can began a type.

**Comment by matklad** [comment]
To update test data, you can run `cargo gen-tests` command.

File: crates/ra_syntax/src/grammar/types.rs [hunk 2]
```diff
@@ -36,6 +36,14 @@ fn type_with_bounds_cond(p: &mut Parser, allow_bounds: bool) {
     }
 }
 
+pub(super) fn is_type_start(p: &mut Parser) -> bool {
```
Comment (matklad) [inline crates/ra_syntax/src/grammar/types.rs]: We probably can use https://github.com/rust-analyzer/rust-analyzer/blob/e396cb3548c8a36ba505cc08248333c225e56d2a/crates/ra_syntax/src/grammar/types.rs#L3 here.

`is_type_start(p)` should be replaced with `p.at_ts(TYPE_FIRST)`.

**Comment by regiontog** [comment]
> 
> 
> To update test data, you can run `cargo gen-tests` command.

For some reason I could not get this to work, I ended up deleting the .txt files that needed updating.

https://github.com/rust-analyzer/rust-analyzer/blob/e396cb3548c8a36ba505cc08248333c225e56d2a/crates/test_utils/src/lib.rs#L266-L273

File: crates/ra_syntax/src/grammar/types.rs (hunk 2 above)
Comment (regiontog) [reply crates/ra_syntax/src/grammar/types.rs]: Wont this fail if the pred is a path starting with `crate` or `super`? I don't see those tokens in `TYPE_FIRST`.

Comment (matklad) [reply crates/ra_syntax/src/grammar/types.rs]: TYPE_FIRST includes PATH_FIRST: `paths::PATH_FIRST.union`. That includes thouse tokens:

https://github.com/rust-analyzer/rust-analyzer/blob/e396cb3548c8a36ba505cc08248333c225e56d2a/crates/ra_syntax/src/grammar/paths.rs#L4

Comment (regiontog) [reply crates/ra_syntax/src/grammar/types.rs]: Aha, ofcourse

**Comment by matklad** [comment]
bors r+

thanks!

**Comment by regiontog** [comment]
Ofcourse, thanks for the help 😄

**Pull request merged**